### PR TITLE
Initialize TCP keepalive and no-delay socket options

### LINKED ISSUE #534: Uninitialized flags in setup_socket can disable TCP_NODELAY and cause ~41 ms TLS SET latency
## Summary

`shard_connection::setup_socket()` passes an uninitialized `int flags` to `setsockopt()` for both `SO_KEEPALIVE` and `TCP_NODELAY`. The later `flags = 1` assignment occurs after those calls.

In our Linux build, a syscall trace showed `TCP_NODELAY=0` on all 128 benchmark connections. Initializing the variable before use and rebuilding removed a repeatable ~41 ms SET latency plateau in our Redis TLS workload.

## Affected source inspected

- Runtime reproduction and local repair: commit `272eeb647df5adc6f67652dfcc177a18dbd88d3e`.
- The same faulty initialization is present in commit `5694a3d61aaf0322a62fc44083ba6f2130b16265` (source inspection only; not benchmarked).
- Source: https://github.com/redis/memtier_benchmark/blob/5694a3d61aaf0322a62fc44083ba6f2130b16265/shard_connection.cpp

The function declares `int flags;`, uses `&flags` for both socket options, and only later assigns it while setting nonblocking behavior. Successful `setsockopt()` return values do not establish that the intended options were enabled.

## Environment and workload

- Linux, separate Redis server and benchmark client over TCP.
- Redis 7.0.15; process restricted to two logical CPUs; `io-threads=2` and `io-threads-do-reads=yes` configured. These are configuration values, not a claim about measured effective TLS threading.
- memtier reports `v=255.255.255 sha=272eeb64:0`, libevent 2.1.12-stable, OpenSSL 3.0.13.
- TLS 1.3, `TLS_AES_128_GCM_SHA256`, no client certificate required.
- Approximately 300,000 preloaded keys, 1,024-byte values, 20% SET / 80% GET.
- 8 threads x 16 connections = 128 persistent connections; pipeline depth 1.
- Isolated lab: server certificate verification was disabled following a certificate-validity error consistent with clock skew. Encryption remained enabled. This is not a production security configuration.

## Reproduction and socket evidence

With an existing TLS Redis endpoint and preloaded keyspace, set `REDIS_HOST` to the test server and run:

```bash
strace -f -e trace=setsockopt -o /tmp/memtier-sockopts.log \
  ./memtier_benchmark --tls --tls-skip-verify \
  -s "$REDIS_HOST" -p 6380 \
  --ratio=1:4 --key-pattern=R:R --key-maximum=300000 \
  --data-size=1024 -t 8 -c 16 --pipeline=1 \
  --test-time=10 --hide-histogram

grep TCP_NODELAY /tmp/memtier-sockopts.log
```

Before the fix, the trace contained 128 successful calls setting the option to zero; representative excerpt:

```text
setsockopt(7, SOL_TCP, TCP_NODELAY, [0], 4) = 0
setsockopt(8, SOL_TCP, TCP_NODELAY, [0], 4) = 0
```

The value of an uninitialized variable is not reliable; other builds may behave differently. The source defect exists regardless of whether a particular binary happens to enable the option.

## Minimal patch tested locally

```diff
diff --git a/shard_connection.cpp b/shard_connection.cpp
--- a/shard_connection.cpp
+++ b/shard_connection.cpp
@@ -392,7 +392,7 @@
 int shard_connection::setup_socket(struct connect_info *addr)
 {
-    int flags;
+    int flags = 1;
     int sockfd;
 
     if (m_unix_sockaddr != NULL) {
         sockfd = socket(AF_UNIX, SOCK_STREAM, 0);
```

A dedicated initialized socket-option variable could also separate these values from the later `fcntl()` flags. The patch above is the minimal change we tested; it initializes the argument for both `SO_KEEPALIVE` and `TCP_NODELAY`.

## Validation and measured results

After rebuilding the same revision, a separate 3-second traced diagnostic passed checks requiring exactly 128 `TCP_NODELAY` calls, all with value `[1]` and successful return status.

We then ran an untraced 60-second comparison using the rebuilt binary and the same command arguments, with `--test-time=60`. Redis was not restarted, and the keyspace was not reloaded between the before/after measurements. No hardware-counter collection was used in either measured run. Traced diagnostic throughput is excluded below.

| Metric | Before repair | After repair |
|---|---:|---:|
| Total ops/s | 15,176.08 | 196,493.13 |
| SET average latency (ms) | 40.92694 | 0.89674 |
| GET average latency (ms) | 0.30115 | 0.58980 |
| Total average latency (ms) | 8.43077 | 0.65119 |
| Total p99 latency (ms) | 41.727 | 1.319 |
| Total p99.9 latency (ms) | 41.983 | 1.575 |

Rebuilt binary SHA-256:
`1e1ffbc73721633c2596938391bd61614d4ea84854e803a7c946fef1bcf65fe7`

Throughput improved by approximately 12.95x in this workload, and the ~41 ms SET plateau disappeared without increasing pipeline depth. These are single-run observations, not a universal speedup or a statistical characterization.

The socket defect and runtime option values are confirmed. The timing and response to the fix strongly support a Nagle/delayed-ACK interaction, but we have not captured packet-level evidence establishing the precise sequence. This is not a TLS-versus-plaintext overhead comparison.

## Suggested regression coverage

Verify with `getsockopt()` that newly created TCP sockets have `TCP_NODELAY=1` and `SO_KEEPALIVE=1`, including reconnect paths. A warning or sanitizer check for uninitialized values would provide additional coverage.

The local rebuild and targeted socket/benchmark checks passed. The full upstream test suite has not been run. Please let us know whether this duplicates an existing report or whether a different fix is preferred.

### PR DESCRIPTION
**Pull request by fcostaoliveira**
TCP connections pass an uninitialized integer to `SO_KEEPALIVE` and `TCP_NODELAY`, so either option can silently remain disabled. Initialize both with a dedicated constant before opening the connection; keep the `fcntl` flags separate. This also applies to TLS and reconnects.

Add a `make check` regression test that inspects real socket options through the production connection setup. It covers IPv4/IPv6, plain/TLS socket setup, Unix nonblocking setup, disconnect cleanup, and two reconnects per configuration. Checks remain active with `NDEBUG`.

Validation:
- Old source compiled with diagnostic `-ftrivial-auto-var-init=zero` fails all 24 TCP option assertions; the fixed source passes. An ordinary unfixed build can happen to pass because an uninitialized integer can be nonzero; this diagnostic makes the regression reproducible.
- Optimized `make check`: 3/3 passing; TLS-disabled, ASAN+UBSan+LSan, TSAN, and real OpenSSL 1.0.2u builds also pass the new test. Every changed executable runtime line is covered (3/3).
- `make format` and `make format-check` pass.
- Full local standalone/cluster × plain/TLS matrix: **896 passed, 468 expected skips, zero failures**. The final standalone TLS run includes test-routing fix #543 (now merged into master) and a private authentication adapter for isolated password-protected fixtures; it is combined validation, not an unmodified checkout run. The adapter preserves explicit credentials and supplies URI credentials inside the URI. An independent CLI/URI authentication cleanup defect found during validation is fixed separately. Original failing logs are retained. macOS CI runs `make check` across four SSL/libevent configurations.

Fixes #534. This verifies socket configuration directly; it does not claim a new measurement of the reporter's TLS latency plateau.

Enabling the new socket tests on macOS exposed a pre-existing CI dependency mismatch: Homebrew libevent used default OpenSSL 3 while memtier selected 1.1 or another 3.0 installation. LLDB showed the 1.1-linked test crashing inside OpenSSL 3's SSL_CTX_free. Build libevent 2.1.13 against each selected SSL and verify transitive library prefixes. All four macOS configurations passed actual unit tests at b00b7e0, including the legacy 1.0.2 build.

The no-TLS CI job runs both normal Prometheus configurations first, then recompiles only the socket test with diagnostic zero initialization. Production objects retain their normal flags while reintroducing the original uninitialized argument fails deterministically.

Release-note text: Fixed uninitialized TCP socket options so keepalive and TCP_NODELAY are consistently enabled, including TLS and reconnects. This can remove artificial latency and change benchmark results compared with affected builds.

The macOS matrix also tests the native Homebrew libevent/default OpenSSL 3 packages. All four macOS variants share the transitive link checker; the known-good legacy libevent 2.1.12 version is retained to avoid an unrelated upgrade.

Updated onto master `830ddbe`. The socket fixture now uses a real client manager and deliberately dispatches Unix connection callbacks on each reconnect with a zero-depth pipeline. Local unit checks and ASAN+LSAN+UBSan with NDEBUG pass; the NULL-manager negative control fails UBSan on that same dispatched callback. The latest CI run checks the combined branch.

---

> [!NOTE]
> **Medium Risk**
> Changes live TCP/TLS connection setup in `shard_connection`, which can alter latency and benchmark numbers; scope is limited to socket option initialization plus tests and CI hardening.
> 
> **Overview**
> Fixes a bug in `shard_connection::setup_socket` where **`SO_KEEPALIVE` and `TCP_NODELAY` were passed an uninitialized `flags` value**, so those options could stay off. Both now use an explicit **`enabled = 1`**; **`fcntl` non-blocking setup** stays on its own `flags` path.
> 
> Adds **`socket_options_test`** to `make check`: it drives the real connection path (IPv4/IPv6, plain/TLS when built, Unix, reconnects) and asserts keepalive, Nagle off, linger, non-blocking, and cleanup. The no-TLS CI job also rebuilds that test with **`-ftrivial-auto-var-init=zero`** so the old bug fails deterministically without changing normal binaries.
> 
> **macOS CI** is tightened: libevent is built or selected to match each OpenSSL matrix entry, **`make check` runs on all macOS jobs**, and **`check-macos-tls-links.sh`** replaces inline `otool` checks so mixed Homebrew SSL/libevent ABIs are caught.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f8f4050b09b5fbc3a806a01889368191dc234942. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by fcostaoliveira** [comment]
Addressed in c3c4831: boolean socket options now require nonzero, supporting BSD/macOS bitmask returns; added SO_LINGER enabled/zero-timeout checks; documented the system-header ordering, main-owned helper stubs and Automake object prefixes.

The regression limitation is real: an uninitialized value can happen to be nonzero, so an ordinary unfixed build is not guaranteed to fail. The old-code diagnostic uses `-ftrivial-auto-var-init=zero` and fails all 24 TCP-option checks; the fixed code passes. The test inspects actual kernel options rather than replacing setsockopt. I will state this explicitly in the PR description.

Keeping real production objects gives UBSan the actual client RTTI and avoids fake types or sanitizer exclusions. A shared library conversion would broaden this socket fix substantially; the object filter is documented and a mismatch fails at link time.

The single NONBLOCK event-loop call drains all active callbacks, including libevent's deferred bufferevent finalizer. In this fixture no connection callbacks are dispatched before disconnect and no additional references are retained. The finalizer closes the descriptor on both socket and TLS paths. This was checked against libevent 2.1.12 source and passes sanitizer, plain/TLS and legacy OpenSSL tests locally.

**Comment by fcostaoliveira** [comment]
Added a committed CI gate for the original uninitialized-read regression: the existing no-TLS job now sets `CXXFLAGS=-g -O2 -ftrivial-auto-var-init=zero` for both Prometheus configurations before running `make check`. This is a test-only build, not a release artifact. The old-code failure is now enforced by CI rather than only described as a local diagnostic.

The macOS dependency change is required to run the new test there: enabling `make check` exposed SIGBUS in both Homebrew-linked variants, and LLDB showed the OpenSSL 1.1-linked executable crashing inside OpenSSL 3's SSL_CTX_free. The existing Homebrew libevent was built against a different SSL installation. The change keeps libevent at 2.1.13 and aligns its dependency; all three macOS jobs now pass. The PR description includes this cause, and the dependency fix is its own commit for review/bisect.

The socket recorder avoids modifying the production API or relying on nonportable descriptor enumeration. Its header-order constraint is documented and the actual test now runs on Linux and macOS SSL variants. The purity audit only covers the unconditional metrics layer (`tests/run_local_gate.sh`); its existing small link is unchanged. The socket test has a separately documented full-object link to retain real UBSan RTTI.

Changed the allocation-failure path to clean up and return, allowing Unix-directory cleanup. There are no other active connection callbacks or threads opening descriptors between disconnect and the EBADF check in this fixture.

I kept existing setsockopt error-handling semantics in this focused runtime fix; changing release-build failure/retry behavior would need its own failure-injection coverage. Added release-note wording to the PR description: enabling TCP_NODELAY consistently can remove artificial latency and change benchmark results compared with affected builds.

**Comment by fcostaoliveira** [comment]
Updated the diagnostic build to keep both normal no-TLS configurations unchanged and then rebuild only the socket test translation unit with zero initialization. This catches the original bug without zeroing unrelated production objects. The metrics-only build comment is now scoped directly to that target.

Restored native packaged coverage with an additional default OpenSSL 3/Homebrew libevent macOS cell. All four macOS variants now use the same transitive link checker. The legacy 1.0.2 job intentionally retains its already-tested libevent 2.1.12 build; the repaired 1.1/3.0 jobs retain 2.1.13, so this repair changes dependency alignment without also upgrading the legacy dependency.

Local normal and diagnostic socket tests pass. Twelve independent checker fixtures cover legacy/native prefixes and foreign/missing dependencies. Final remote checks are running. Full local validation results will be posted before calling this ready.

**Comment by fcostaoliveira** [comment]
Removed continue-on-error from the macOS job, so its unit and linkage checks now gate CI. The preceding exact-head run already passed all four variants, including native Homebrew (3): https://github.com/redis/memtier_benchmark/actions/runs/35588389383 . That job's actual otool output and successful prefix check confirm the opt-prefix match; the legacy job also passes the shared prefix checker.

The diagnostic step now passes only SOCKET_TEST_EXTRA_CXXFLAGS; its normal flags remain defined once in Makefile.am. Detection in configure is unnecessary for this explicit diagnostic step on the known ubuntu-latest compiler, and normal builds retain no extra flag. Local normal and extra-flag checks pass.

The matrix values select actual OpenSSL installations: 3.0 is the pinned minor formula and 3 follows Homebrew's current default. The additional package/custom choice follows from whether Homebrew's libevent is built against that SSL. The legacy source-build recipe intentionally stays at its prior version; consolidation or upgrading it is not needed to repair the dependency mismatch.

The recorder/real-object test design and deferred cleanup were reviewed independently. NONBLOCK drains active callbacks until none remain in libevent2.1.12; this fixture has no retained extra bufferevent references or connection callbacks opening descriptors. Adding a retry would hide a cleanup regression rather than fix one. The main-owned helper stubs remain local to avoid a production-library refactor in a socket option fix.

File: tests/unit/socket_options_test.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                check_option(fd, IPPROTO_TCP, TCP_NODELAY, label, "TCP_NODELAY");
+                struct linger linger_value = {};
+                socklen_t len = sizeof(linger_value);
+                if (check(getsockopt(fd, SOL_SOCKET, SO_LINGER, &linger_value, &len) == 0, label, "SO_LINGER")) {
+                    check(linger_value.l_onoff != 0 && linger_value.l_linger == 0, label, "abortive close enabled");
+                }
+            }
+            connection.disconnect();
+            // libevent can defer bufferevent destruction until the loop runs.
+            event_base_loop(base, EVLOOP_NONBLOCK);
```
[the remark this answers is not in the document]
Comment (fcostaoliveira) [reply tests/unit/socket_options_test.cpp:158]: Addressed in 388d3cc, now included in b944660 on latest master830ddbe. The fixture owns a real client manager, and it explicitly dispatches the Unix connection callback on every reconnect with a zero-depth pipeline, verifying connected state and zero generated requests.

The immediate-success path is deferred in the libevent versions used by CI, and bufferevent_free clears callbacks before the cleanup drain. However, the advertised older libevent2.0 hard-error path can invoke a callback synchronously, so removing the NULL-manager assumption is worthwhile.

Validation: makecheck3/3, formatting, and ASAN+LSAN+UBSan with NDEBUG all pass. As a negative control, the old NULL manager with the same explicit Unix callback dispatch fails UBSan on the manager dereference; the real-manager fixture passes. Independent review approved the ownership and cleanup. Unit checks also passed after merging latest master. CI is running on the updated PR head.

**Comment by fcostaoliveira** [comment]
Final local suite results are now complete and included in the PR description: 896 passed, 468 expected skips, zero failures across standalone/cluster and plaintext/TLS. The final standalone TLS run included the now-merged #543 test-routing fix and a private adapter for authenticated fixtures; original failing logs were retained. This PR now contains master830ddbe directly, and makecheck3/3 passed after that merge.

The updated real-client socket fixture also passed 500 consecutive optimized runs locally (4.02 seconds), plus ASAN+LSAN+UBSan with NDEBUG. Its Unix connected callback is exercised explicitly, with no Redis requests generated. The obsolete NULL-manager assumption is removed.

The macOS CI work is retained because executing this regression test exposed the existing mixed-OpenSSL libevent crash, so aligning those dependencies is needed to run the test across the supported SSL variants. Making these checks mandatory is deliberate. The native Homebrew cell preserves packaged-dependency coverage; the existing legacy recipe stays at its known-good version. The repository has no committed changelog; the release-note text remains in the PR description for the repository's curated release workflow.

The recorder remains limited to the socket test translation unit, with socket headers included before macro substitution. This keeps real socket options and production connection setup under test across both GNU and macOS linkers without adding a production testing interface. Current-head CI and bot review are running.

**Comment by fcostaoliveira** [comment]
Addressed the OpenSSL compatibility and optional-flag declaration points in f8f4050. The test now mirrors production's pre/post-1.1 initialization and method selection. Modern TLS, actual OpenSSL1.0.2u with matching libevent, and TLS-disabled builds each pass makecheck3/3; formatting and independent review pass. The explicit empty flag default remains overrideable from make's command line.

The fixture's coverage claim is deliberately socket setup/options and reconnect descriptor cleanup. A zero-depth pipeline prevents sending Redis requests to bare socket listeners while safely exercising the real Unix connection callback. Full production connection/protocol behavior is covered by the integration suites; this unit test does not claim a TLS handshake against its bare listeners.

The macOS gating change is deliberate under the requested cross-variation validation. Dependency fetches already retry three times and verify the pinned archive checksum; a persistent fetch or linkage failure should remain visible. Keeping the dependency repair here lets the new test actually run on the SSL variants that exposed the mismatch. This is also covered by independent review.

The test-only recorder was chosen to inspect the production socket without adding a public production accessor or a platform-specific linker wrapper. System socket declarations are included before macro substitution. Compilation across the supported toolchains makes header changes that conflict with the recorder visible immediately; a production testing interface would broaden this repair unnecessarily. The earlier reply and updated description include the full-suite results, 500-run fixture check, and release-note rationale.

**Comment by fcostaoliveira** [comment]
The macOS failure policy is intentional, not an accidental consequence of adding makecheck. These four configurations now validate the actual unit tests and transitive SSL linkage, and the native Homebrew leg retains package coverage. The dependency alignment is required by the crashes uncovered when enabling this test. I am retaining that repair together with its cross-platform validation.

config_init_defaults is static in the executable's main translation unit. Exporting it or linking main into this test would require a broader refactor. More importantly, applying its real defaults and then overriding them still requires pipeline0 here: these are bare socket listeners, so the fixture must not send a Redis workload. The test explicitly claims socket-option setup and reconnect cleanup, while production CLI/default behavior stays in the full integration suites. The real client manager is used so dispatched connection callbacks have correct ownership and statistics.

Extracting only the setsockopt calls would let a unit test pass even if production connect stopped invoking that helper. The current recorder tests the actual connection path and a fresh descriptor on reconnect, with real getsockopt checks. The test-only stubs are intentionally minimal; the shared fuzz stubs define object-generator implementations that conflict with the real production objects needed here. Moving process-wide helpers is a separate refactor, not necessary for this regression fix. Independent reviewers approved these choices; supported builds and current-head CI exercise their portability.

**Comment by fcostaoliveira** [comment]
Final verification on f8f4050b09b5fbc3a806a01889368191dc234942: all applicable CI is green, 120 successful checks and6 intentional skips. This includes the four macOS variants, eight RPM builds, full plain/TLS integration suites, sanitizer suites, coverage, Claude and Cursor. All substantive bot findings have been addressed; the Cursor thread is resolved. Independent code review approved this exact head.

The branch includes latest master830ddbe. Local evidence remains documented above: 896full-suitepasses/468expectedskips,3/3changed runtime lines covered, normal/no-TLS/legacySSL unit builds, sanitizer checks, and500consecutive socket fixture runs. No merge is performed by this update.
